[ERE-1547] Report Redesign

ppettitau opened the pull request:
Redeveloped Reporting for TRRF
Feature switch to require explicit enabling of legacy reporting functionality

ppettitau committed:
Redeveloped Reporting for TRRF
Feature switch to require explicit enabling of legacy reporting functionality

ppettitau committed:
Fix failing checks

ScriptSmith commented on rdrf/report/views.py:
```diff
@@ -0,0 +1,92 @@
+        return render(request, 'report_designer.html', params)
+
+    def post(self, request, report_id=None):
+
```
nit: empty line

ScriptSmith commented on rdrf/report/views.py:
```diff
@@ -0,0 +1,92 @@
+
+class ReportDesignView(SuperuserRequiredMixin, View):
+    def get(self, request, report_id=None):
+
```
nit: empty line

ScriptSmith commented on rdrf/explorer/views.py:
```diff
@@ -43,11 +49,13 @@ class NewQueryView(SuperuserRequiredMixin, View):
 
     @csp_update(SCRIPT_SRC=["'unsafe-eval'"])
     def get(self, request):
+        check_legacy_reports_enabled()
```
These checks could be decorators

ScriptSmith commented on rdrf/rdrf/schema/schema.py:
```diff
@@ -0,0 +1,236 @@
+        context_form_ids = clinical_data.values_list("context_id", flat=True)
+        context_lookup = {context.id: context for context in (RDRFContext.objects.filter(pk__in=list(context_form_ids)))}
+
+        cfg_data_cnt_lookup = {context.context_form_group.id: 1 for context in RDRFContext.objects.filter(pk__in=list(context_form_ids))}
```
No need to perform the same db query twice

ScriptSmith commented on rdrf/rdrf/services/io/defs/exporter.py:
```diff
@@ -521,6 +525,18 @@ def _get_reports(self):
+                 'clinical_data_fields': [f.cde_key for f in r.reportclinicaldatafield_set.all()],
+                 'demographic_fields': [{'model': f.model, 'field': f.field, 'sort_order': f.sort_order} for f in r.reportdemographicfield_set.all()]
+                 }
+                for r in (ReportDesign.objects.filter(registry=self.registry))]
```
Possible to use select_related / prefetch_related here?

ppettitau replied:
Like this?

```
    def _get_reports_v2(self):
        return [{'title': r.title,
                 'description': r.description,
                 'registry': r.registry.code,
                 'access_groups': [ag.name for ag in r.access_groups.all()],
                 'filter_working_groups': [wg.name for wg in r.filter_working_groups.all()],
                 'filter_consents': [{'section': c.section.code, 'code': c.code} for c in r.filter_consents.all()],
                 'clinical_data_fields': [f.cde_key for f in r.reportclinicaldatafield_set.all()],
                 'demographic_fields': [{'model': f.model, 'field': f.field, 'sort_order': f.sort_order} for f in r.reportdemographicfield_set.all()]
                 }
                for r in (ReportDesign.objects.filter(registry=self.registry)
                                              .select_related('registry')
                                              .prefetch_related('access_groups',
                                                                'filter_working_groups',
                                                                'filter_consents',
                                                                'reportclinicaldatafield_set',
                                                                'reportdemographicfield_set'))]
```

ppettitau replied:
@ScriptSmith , see above comment.

ScriptSmith replied:
Yep sorry. Looks like it's working correctly.

ScriptSmith commented on rdrf/rdrf/settings.py:
```diff
@@ -345,7 +348,7 @@
 CSP_OBJECT_SRC = ["'none'"]
 CSP_SCRIPT_SRC = get_csp(
-    ["'self'", "'unsafe-inline'"],
+    ["'self'", "'unsafe-inline'", "'unsafe-eval'"],
```
This should be allowed only in specific views, possibly using a decorator like so:

```
@csp_update(SCRIPT_SRC=["'unsafe-eval'"])
```

ppettitau replied:
This is required for the graphiql view. As this is provided by the Graphene-django library, I'm unsure how else to configure it.

ScriptSmith commented on rdrf/rdrf/templates/rdrf_cdes/report_deprecation_warning.html:
```diff
@@ -0,0 +1,6 @@
+<div class="alert alert-warning">
+    <p>{% trans 'The TRRF Reporting Tool has now been deprecated.' %}</p>
+    <p>{% trans 'We suggest checking out the newly redeveloped Reports module with enhanced user interface and reporting capabilities.' %}</p>
+    <p><a href="{% url 'report:reports_list' %}">{%  trans 'Visit the new Reports module' %}</a></p>
```
nit: two spaces

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -0,0 +1,252 @@
+            ReportCdeHeadingFormat.ABBR_NAME.value: ('cfg.abbreviatedName', 'form.abbreviatedName', 'section.abbreviatedName', 'cde.abbreviatedName'),
+            ReportCdeHeadingFormat.LABEL.value: ('cfg.name', 'form.name', 'section.name', 'cde.name'),
+            ReportCdeHeadingFormat.CODE.value: ('cfg.code', 'form.name', 'section.code', 'cde.code')
+        }[self.report_design.cde_heading_format]
```
nit: Creating a dict for this purpose is a bit unusual, it's more concise with less repetition, but will be slower - not that it really matters here.

I think it's better as an if statement (we'll get the `match` statement after #386 is merged), or creating a helper method.

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -0,0 +1,252 @@
+            return pandas_field
+
+        def raise_pivot_exception(e):
+            if self.report_design.cde_heading_format == ReportCdeHeadingFormat.ABBR_NAME.value \
```
This same exception can occur when using a `LABEL` format as well. Additionally, this error needs to be explained to the user (currently they will just see `Server error`), and should guide them on potential fixes, along with their caveats.

Might be better to pre-compute any potential duplicates so they can all be fixed at once, not one-by-one.

ppettitau replied:
Something like this?
![image](https://user-images.githubusercontent.com/60163872/153103409-8e1c9916-7a09-487c-a87f-5d9fd410ebd9.png)

ScriptSmith replied:
Perfect :+1:

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -0,0 +1,252 @@
+
+    def export_to_json(self, request):
+        result = schema.execute(self.__get_graphql_query(), context_value=request)
+        return json.dumps(result.data)
```
This will need to work when the size of the data exceeds available memory on the server (same with the CSV formatting).

See if there's a way to get the graphql response in the form of an generator / iterable, that you can return in a `StreamingHttpResponse`, as [Line-delimited JSON](https://en.wikipedia.org/wiki/JSON_streaming#Line-delimited_JSON) rather than JSON

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -0,0 +1,252 @@
+        cde_pivot_cols = [cfg_name_col, 'cfg.sortOrder', 'cfg.entryNum', form_name_col, section_name_col, 'section.entryNum', cde_name_col]
+
+        try:
+            pivoted = df.pivot(index=demographic_cols,
```
The `cfg.code` column is missing when I pivot using `CODE` report type:

> KeyError: 'cfg.code'

ScriptSmith reviewed (commented):
Looking good from what I've seen so far. Will probably need a week to do a full review, but have left some initial comments.

ScriptSmith commented on rdrf/rdrf/schema/schema.py:
```diff
@@ -0,0 +1,236 @@
+import logging
```
Any reason this file can't be moved into the `report` app?

ScriptSmith commented on rdrf/rdrf/schema/schema.py:
```diff
@@ -0,0 +1,236 @@
+    def resolve_all_patients(self, info, registry_code, filters=[], working_group_ids=[]):
+        registry = Registry.objects.get(code=registry_code)
+
+        query_args = dict([query_filter.split('=') for query_filter in filters])
```
Can we make all the query filters part of the query fields, rather than accepting anything?

ppettitau committed:
Update default values for abbreviated name to use equivalent 'name' fields from the source models

ppettitau committed:
Add report unit tests and code modifications for bug fixes discovered during testing.

ppettitau committed:
Update report deprecation warning

ppettitau committed:
Fix failing test

ppettitau committed:
Fixes from PR Feedback, including:
* fixing erroneous whitespace
* move schema from rdrf app into report app
* pre-load related data when exporting reports_v2 data
* pre-validate report_design configuration for duplicate headings before CSV export
* Convert legacy reports check into a decorator
* Move graphiQL to only be accessible in DEBUG mode

ppettitau committed:
Merge branch 'next_release' into ERE-1547_Report_Redesign

ppettitau committed:
Modify graphQL schema to support basic pagination of patient list
Modify json export to stream patient data in chunks of 20

ppettitau commented on rdrf/report/reports/generator.py:
```diff
@@ -0,0 +1,282 @@
+    def __init_report_fields_lookup(self):
+        return {model: model_config['fields'] for model, model_config in self.report_config.items()}
+
+    def __humanise_column_label(self, col):
```
Not working where pivot model field is not simple e.g. `address_type { type }`

ppettitau committed:
Stream CSV data in Http Response
Fix bug with humanising complex pivoted data (e.g. addressType { type })

ppettitau committed:
Add ability for registries to extend reporting

ppettitau committed:
Merge branch 'next_release' into ERE-1547_Report_Redesign

ScriptSmith committed:
Merge branch 'next_release' into ERE-1547_Report_Redesign

ScriptSmith committed:
Create schema dynamically

ScriptSmith committed:
Add docs to schema

ScriptSmith committed:
Respond to GraphQL requests by creating the schema lazily

Required because schema may change after registry structure is changed,
and to support running django migrations when the database hasn't been
fully configured

ppettitau committed:
* Refactor CSV Export to pre-calculate headers and use DictWriter to stream CSV to client
* Rollback deletion of some DjangoObjectTypes in graphql schema that are still required
* Add initial implementation of data summary type to graphql schema
* Remove pandas dependency

ppettitau committed:
Merge branch 'next_release' into ERE-1547_Report_Redesign

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
+                context_lookup = {context.id: context for context in patient_contexts}
+                cfg_counter_lookup = {context.context_form_group.id: 0 for context in patient_contexts}
+
+                for idx, entry in enumerate(clinical_data):
```
`enumerate` not necessary

ppettitau replied:
Fixed.

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
+
+                    cnt_cde_values = len(cde_value) if type(cde_value) is list else 1
+                    max_cnt_cde_values = max(cnt_cde_values, cd_headers.get(cde_key, {"count": 1})['count'])
+                    cd_headers[cde_key] = {"header": cde_header, "count": max_cnt_cde_values}
```
This will be quite slow because:

1. The `cde_headers` dict will be assigned to each time
2. The key string needs to be built each time
3. It's being run for every value in the clinical data

You can fix this with a conditional update, memoization / prebuilding the list of keys, and filtering the values to count according to the query.

But is it worth making the datastructure hierarchical with codes when doing the counting, rather than flat with partially flattened string keys? Way less complexity due to the string building that way, and you can build each formatted heading once right at the end, rather than each time.

ppettitau replied:
Can you provide an example of the datastructure you're suggesting?

ScriptSmith replied:
Something like:

```python
{
    "longitudinal_cde_code": {
        "count": 4,
        "forms": {
            "form_name": {
                "multi_section_code": {
                    "count": 2,
                    "cdes": {
                        "multi_value_cde": {
                            "count": 8
                        }
                    }
                }
            }
        }
    }
}
```

Another way could be keeping it flat, but using tuples as keys instead of strings. Eg:

```python
cde_key = (cfg, cfg_num, form, section, section_num, cde)
if cnt_cde_values > cde_headers.get(cde_key, 0): # Could make cde_headers a default dict?
    cd_headers[cde_key] = cnt_cde_values
```

And as we mentioned the other day, it is possible to calculate all this on the database end:

- ` jsonb_array_length` in https://www.postgresql.org/docs/current/functions-json.html#FUNCTIONS-JSON-PROCESSING-TABLE
- https://gist.github.com/alej0varas/4d66cfd46e96d8c6a0bb5513afa3664c

That would probably be significantly faster, but will be complex to generate the query for each thing you need to count

ppettitau replied:
Thanks. I'll play around and see what might be the easiest way to go.

ScriptSmith commented on rdrf/report/schema/schema.py:
```diff
@@ -212,6 +267,28 @@ def clinical_data_resolver(patient, _info):
 # TODO: memoize + possible cache clearing when registry definition changes?
 # TODO: Replace partial resolvers with single resolve function for each level
 # TODO: Replace Metaprogramming with a low-level library like graphql-core
+def list_patients_query(user,
```
The TODO comments need to be above their original function

ppettitau replied:
Fixed

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
+                if cde_cnt == 1:
+                    flat_cd_headers[key] = header
+                else:
+                    for i in range(0, cde_cnt):
```
The `0` is implied if you only provide the limit.

ppettitau replied:
Fixed

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
-
-        return stream_to_csv(pivoted)
+        def get_demographic_headers():
+            def get_flat_json_path(report_model, report_field, i=None):
```
I generally prefer [explicit names](https://peps.python.org/pep-0020/) for variables, rather than abbreviations and single-character names. (acronyms are ok if they're very verbose otherwise and referenced often, I also sometimes make an exception for simple comprehensions)

It makes things less ambiguous when reading & writing, and makes you less likely to accidentally reference or assign the value in a different scope.

ppettitau replied:
Fixed

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
+                        for section_data in form_data.get("sections", []):
+                            section = next(s for s in sections if s.code == section_data['code'])
+                            cdes = section.cde_models
+                            for i, cde_data in enumerate(section_data.get("cdes", [])):
```
See explicit naming comment

ScriptSmith commented on rdrf/report/schema/schema.py:
```diff
@@ -227,22 +304,15 @@ def resolve_patients(_parent,
+        return list_patients_query(info.context.user, registry_code, consent_question_codes, working_group_ids)[offset:limit]
 
-        return patient_query.distinct()[offset:limit]
+    def resolve_data_summary(_parent, info, registry_code, consent_question_codes=[], working_group_ids=[]):
```
Can't use empty lists as default parameters: https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
+            return flat_cd_headers
+
+        # Build Headers
+        cde_keys = [rcdf.cde_key for rcdf in self.report_design.reportclinicaldatafield_set.all().order_by('id')]
```
Can use `values_list("cde_key", flat=True)`

ppettitau replied:
Fixed

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
+
+            return fieldnames_dict
+
+        def get_clinical_data_headers(cde_keys):
```
This will probably benefit from being moved into a separate file. This class is getting large, and could end up being another mega-class like we see in other parts of the codebase. Worth trying to split it up and write it functionally?

ppettitau replied:
Yes, I agree. I might need some pointers with how to write it functionally given my very OO-heavy background :) 
I'll get back to you.

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
+                            model_fields = self.report_design.reportdemographicfield_set.filter(
+                                model=rdf.model).values_list("field", flat=True)
+
+                            for i in range(0, (num_variants or 0)):
```
The 0 is implied if you only provide the limit.

ppettitau replied:
Fixed

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
+                    prefix = f'{report_model}_'
+
+                field_json_path = re.sub(r"\s", "", report_field)
+                field_json_path = re.sub(r"(.+)({(.*)})", r"\1_\3", field_json_path)
```
Needs explanation

ppettitau replied:
Fixed. Hopefully I haven't over-explained it now.

ScriptSmith commented on rdrf/report/reports/generator.py:
```diff
@@ -197,141 +153,218 @@ def export_to_json(self, request):
+                return summary_result.data['dataSummary'][lookup_key]
+
+            fieldnames_dict = OrderedDict()
+            fieldnames_dict['id'] = 'ID'
```
Worth setting `auto_camelcase=False` in the schema?

https://docs.graphene-python.org/en/latest/api/#schema

ppettitau committed:
* Small PR fixes to generator
* Move comments in schema.py back to their correct location

ppettitau committed:
* Rewrite csv header generator for clinical data (plus unit test)
* Small fix to label for demographic headers to use 1-index based numbering for display
* Fix default values for parameters in schema to use None instead of empty list.

ppettitau committed:
Refactor data summary node to perform aggregation only when that specific data is requested.

ppettitau committed:
* Add initial support for reporting on Registered Clinicians
* Fix demographic csv headings to use format: (model)_(variant index)_(fieldname)

ppettitau committed:
Add xray subsegments to diagnose performance issues

ppettitau committed:
Fix xray subsegments

ppettitau committed:
Fix linting

ppettitau committed:
Use paginator to iterate through clinical data in attempt to resolve memory issues with large datasets.

ScriptSmith committed:
Change outer subsegment to segment

ppettitau committed:
Retrieve clinical data summary via SQL

Apply a basic ordering to clinical data (grouped by form, section)

ppettitau committed:
Filter clinical data summary by patients relevant to a given report (as per user defined filters)

ppettitau committed:
Fix linting issues

ppettitau committed:
Merge branch 'next_release' into ERE-1547_Report_Redesign

ppettitau committed:
Allow for patient demographic fields on the schema to be customised

ppettitau committed:
Add support for CDEs that require denormalising (e.g. Mims in MND)

ppettitau committed:
Merge branch 'next_release' into ERE-1547_Report_Redesign

ppettitau committed:
Fix missed rename of settings

ppettitau committed:
Rename private methods to indicate intended scope

ppettitau committed:
Turn the Reports list table into a DataTable to enable column ordering, pagination and searching for enhanced UX

ScriptSmith commented on rdrf/rdrf/migrations/0140_auto_20220114_1530.py:
```diff
@@ -0,0 +1,106 @@
+# Generated by Django 2.2.24 on 2022-01-14 15:30
```
Migration files should have a slug name rather than 'auto'

ppettitau replied:
Fixed.

ScriptSmith commented on rdrf/report/tests/models_tests.py:
```diff
@@ -0,0 +1,57 @@
+class ReportDesignTestCase(TestCase):
+
+    def get_reports_for_user(self, user):
+        return list(ReportDesign.objects.reports_for_user(user).all())
```
`.all()` is redundant

ppettitau replied:
Fixed.

ScriptSmith commented on rdrf/report/views.py:
```diff
@@ -0,0 +1,93 @@
+from django.shortcuts import render, get_object_or_404, redirect
+from django.views import View
+
+from rdrf import settings
```
Should import from `django.conf`

ppettitau replied:
I don't think I even need the setting that it's referencing, just something I copied over from explorer. Will check.

ppettitau replied:
Fixed - removed the need to import settings.

ScriptSmith commented on rdrf/report/forms.py:
```diff
@@ -0,0 +1,187 @@
+
+    def clean(self):
+        super(ReportDesignerForm, self).clean()
+        cleaned_title = self.cleaned_data['title']
```
Fails if title is missing

ppettitau replied:
Weird because the field should be marked as required at the browser side and validated before it's POSTed. Will check why this isn't happening.

ppettitau replied:
Updated template to use the form field (to render the required attribute as I was expecting). Have also updated this line to safely reference 'title' via get + an empty string default.

ScriptSmith commented on rdrf/report/forms.py:
```diff
@@ -0,0 +1,187 @@
+        cleaned_title = self.cleaned_data['title']
+        if not self.instance.id or self.instance.title != cleaned_title:
+            if ReportDesign.objects.filter(registry=self.cleaned_data['registry'], title=cleaned_title).first():
+                self.add_error('title', _(f'A report in this registry with the title "{cleaned_title}" already exists.'))
```
Possible to just make the model field `unique=True` and ModelForm will create this error automatically?

ppettitau replied:
Not quite. Title is only required to be unique per registry. You could have reports with the same title across different registries.

ScriptSmith commented on rdrf/report/reports/clinical_data_report_util.py:
```diff
@@ -0,0 +1,236 @@
+from django.db import connections
+from django.db.models import Count
+
+from rdrf import settings
```
Should use `django.conf`

ScriptSmith commented on rdrf/report/forms.py:
```diff
@@ -0,0 +1,187 @@
+class ReportDesignerForm(ModelForm):
+
+    registry = ModelChoiceField(label=_('Registry'), widget=Select(attrs={'class': 'form-select'}), queryset=Registry.objects.all(), to_field_name='code')
+    demographic_fields = MultipleChoiceField(label=_('Demographic Fields'), widget=SelectMultiple(attrs={'class': 'form-select', 'size': '10'}))
```
`required=False` too?

ppettitau replied:
Yes, I think so. Previously there was a limitation (perhaps in how I was doing things in Pandas) that required demographic fields. I don't think it's an issue anymore, so I'll look at making it optional.

ScriptSmith commented on rdrf/report/reports/clinical_data_report_util.py:
```diff
@@ -0,0 +1,236 @@
+            for row in cursor.fetchall():
+                form_name, section_code, max_cnt_sections, cde_code, max_cde_value_cnt = row
+                section_cdes = summary.setdefault(form_name, {}).setdefault(section_code, {'count': max_cnt_sections, 'cdes': {}}).get('cdes')
+                section_cdes.setdefault(cde_code, {'count': max_cde_value_cnt})
```
fyi: I think `itertools.groupby` would also work here

And I think the last `setdefault` rather than `[cde_code]` isn't necessary, unless the query output requires it for some reason?

ppettitau replied:
This works:

```
            summary = {
                form_name: {
                    section_code: {
                        'count': section_count,
                        'cdes': {row[3]: {'count': row[4]}
                                 for row in cdes}}}
            for form_name, sections in itertools.groupby(rows, lambda x: x[0])
            for (section_code, section_count), cdes in itertools.groupby(sections, lambda x: (x[1], x[2]))
            }
```

Happy with that?

ScriptSmith replied:
Yeah nice. `groupby` relies on the keys in the iterable being in sorted order, is that guaranteed in the query?

ppettitau replied:
Yes, it will be once I make my changes.

adding this line to the query:
```
ORDER BY cs.form_name, cs.section_code, cs.cde_code;
```

ScriptSmith commented on rdrf/report/schema/schema.py:
```diff
@@ -0,0 +1,386 @@
+import logging
```
nit: this file probably doesn't need its own package

ScriptSmith commented on rdrf/report/reports/clinical_data_report_util.py:
```diff
@@ -0,0 +1,236 @@
+        return max_cnt_entries.first()['cnt_entries'] if max_cnt_entries else 1
+
+    def __get_cfg_data(self, cfg, cfgs_lookup, heading_format):
+        cfg_lookup = cfgs_lookup.get(cfg.code, None)
```
`setdefault` should work here

ppettitau replied:
I believe `setdefault` evaluates the `default` logic even if the key exists in the dictionary. This was my way of avoiding that for the sake of performance. I'll add a comment.

Happy to be proven wrong.

ScriptSmith replied:
Yep, good point. No comment necessary I'd say.

ScriptSmith commented on rdrf/report/reports/clinical_data_report_util.py:
```diff
@@ -0,0 +1,236 @@
+        for cde_i, cde_key in enumerate(cde_keys):
+            form_name, section_code, cde_code = get_form_section_code(cde_key)
+
+            sort_order_lookup.setdefault(form_name, {'order': len(sort_order_lookup), 'sections': {}})
```
Since all the `order` values are just the lengths of the list when the value is set, can this be simplified to use `OrderedDict`s all the way down, and removing all the `__sorted` methods?

ppettitau replied:
I'm not really sure how to do this. 
The purpose of this sort method is to influence the headers to be sorted by the order of cde_keys as they were selected in the report designer UI. 
This is independent to the data summary dict that is built from SQL.

ScriptSmith replied:
Ah true, looking up the 'insert order' in `sorted_order_lookup` would be more complicated :+1:

ScriptSmith commented on rdrf/rdrf/settings.py:
```diff
@@ -727,3 +735,12 @@
+    CSP_EXCLUDE_URL_PREFIXES = '{0}/graphql'.format(SCRIPT_NAME)
+
+GRAPHENE = {
+    "SCHEMA": "report.schema.schema.schema",
```
This line should no longer be necessary

ScriptSmith commented on rdrf/report/tests/clinical_data_generator_tests.py:
```diff
@@ -0,0 +1,260 @@
+
+        report_design.cde_heading_format = ReportCdeHeadingFormat.CODE.value
+        expected = OrderedDict()
+        expected.update({
```
Can pass the dict to the constructor

ScriptSmith commented on rdrf/report/admin.py:
```diff
@@ -0,0 +1 @@
+# Register your models here.
```
Delete

ppettitau replied:
Done.

ScriptSmith commented on rdrf/report/models.py:
```diff
@@ -0,0 +1,76 @@
+
+    model = models.CharField(max_length=255)
+    field = models.CharField(max_length=255)
+    sort_order = models.PositiveIntegerField(null=False, blank=False)
```
Keeping this in the model only preserves the sort order according to the report configuration at the time, as it's not currently customizable in the form. Worth removing it and just using the configuration?

ppettitau replied:
I'm on the fence about this one. 

Pro sort-order:
* Once sort order is determined, it is seamlessly applied through the model Meta 'ordering' attribute without having to explicitly order the result set when it's used
* Allowance for introduction of a sort order for demographic fields in the future
* If I remove it I'll have to write logic to lookup the index of the field in the report_configuration to be run every time we generate the report. 

Cons:
* Not consistent with clinical data (but to be fair, the ordering of clinical data is much more complex)
* Not customisable in the form

I'm thinking it's not worth the effort of changing it but happy to discuss/consider further.

ScriptSmith replied:
Let's keep it for now. Is adding custom ordering documented in the ticket somewhere?

ppettitau replied:
It is now :)

https://eresearchqut.atlassian.net/browse/ERE-1724

ScriptSmith commented on rdrf/report/views.py:
```diff
@@ -0,0 +1,93 @@
+
+
+class ReportDeleteView(SuperuserRequiredMixin, View):
+    def get(self, request, report_id):
```
Move to a `delete` method in `ReportDesignView`?

ppettitau replied:
I would have to change the way the delete works through the UI to use either a form or an ajax request (which is fine, I can look into it), as I'm currently using a link (and you can't pass the HTTP method with a link).
Probably more correct to use a form/ajax anyway.

ppettitau replied:
I think deleting through the link was a bit of lazy programming.

ScriptSmith replied:
All good, link is fine. That's how it's done in other django forms.

ScriptSmith commented on rdrf/report/views.py:
```diff
@@ -0,0 +1,93 @@
+
+
+class ReportsView(View):
+    def get(self, request):
```
It's simpler to implement this way, but my preference would be `PermissionDenied` rather than an empty table if the user isn't supposed to have access to reports. That way, the boundary between patient & staff users is enforced and easily visible at the view-level, rather than in its response logic.

ppettitau replied:
My view is that reports can target Patients, therefore they should be able to see the page. 

It might be confusing for the user if they expected to be able to access the reports list page (had there been a report available to them), but then they get Permission Denied if that specific report got taken away. 

Otherwise if you want me to change it I need to add another check to determine whether to show Permission Denied. 
i.e.
```
if user has no reports AND is not a superuser
   raise PermissionDenied
```

Let me know if you'd like me to proceed with the above.

ScriptSmith replied:
In most of trrf, there is clear separation between admin / staff functionality and patient functionality (admin & explorer pages), that is enforced at the 'response status' level. Previously, this enforcement was handled differently in different parts of the application (often mixed into the response logic), which lead to unexpected security vulnerabilities because it wasn't easy for devs to evaluate whether a view had the correct permissions unless you inspected the application logic.

In this PR, it is currently not possible for patients to access any reports anyway, because `reports_for_user` still requires that the user is a superuser, curator, or clinician. The reports page is also not visible to patients in the quicklinks.

It's possible we could allow patients and carers to have access to reports on themselves, but that would require a good usecase and more testing.

The `PermissionDenied` in the mixin wouldn't check whether any reports have been made available to them, just that they're the type of user that is expected to have access to reports.

ppettitau replied:
Ok, in that case, I will remove those groups from the UI as well.

![image](https://user-images.githubusercontent.com/60163872/161471052-4d02ac5d-962f-4f13-8c9e-73c55aa5e73c.png)

ppettitau replied:
Done.

ScriptSmith replied:
I think that change allows users that aren't in the desired access group to download a report in `ReportDownloadView`.

ppettitau replied:
oops. I've split it into 2 different checks now.

ScriptSmith commented on rdrf/report/forms.py:
```diff
@@ -0,0 +1,187 @@
+class ReportDesignerForm(ModelForm):
+
+    registry = ModelChoiceField(label=_('Registry'), widget=Select(attrs={'class': 'form-select'}), queryset=Registry.objects.all(), to_field_name='code')
+    demographic_fields = MultipleChoiceField(label=_('Demographic Fields'), widget=SelectMultiple(attrs={'class': 'form-select', 'size': '10'}))
```
Possible to use `ModelMultipleChoiceField` for some of the `MultipleChoiceField`s to make them a bit simpler?

ppettitau replied:
Still need to investigate.

ppettitau replied:
I've spent some time looking at this and there's not much I can do with the current implementation to simplify the form logic.

The complexity comes in where I've used json as field values to support the UI in how it conditionally displays options (working groups, consents, CDE fields) based on the registry selected.

ScriptSmith commented on rdrf/explorer/views.py:
```diff
@@ -1,38 +1,38 @@
 
 class MainView(View):
 
+    @is_legacy_reports_enabled
```
`rpc_reporting_command` in `rdrf/rdrf/services/rpc/rpc_commands.py` should also be restricted unless legacy reports are enabled

ppettitau replied:
Still to do.

ppettitau replied:
Done.

ppettitau committed:
Action PR:
* Rename migration to use slug name in preference over default 'auto' name
* Remove schema package and relocate to report package
* Fix settings imports to reference django.conf
* Use itertools.groupby to extract clinical data summary from cursor for increased clarity
* Update demographic fields in Report Designer to be optional
* Update report to no longer force patient ID to be exported with results
* Update report deletion to use an AJAX Delete request

ppettitau committed:
* Restrict report access to only users who are super users, clinicians or curators

* Disable rpc call to explorer if legacy reports feature is not enabled.

ppettitau committed:
* Fix report access
* Fix failed linting and unit tests

ppettitau committed:
* Fix failing unit test

ppettitau committed:
* Update zips
